### ISSUE TITLE
Reminder: Wine Instructions Update for C2VectParallel fix (winetricks)

### ISSUE THREAD

**Issue by RadWolfie** [Cxbx-Reloaded/Cxbx-Reloaded#2161]
## Quick summary
As of Wine 6.3 dev build, C2VectParallel function has been implemented to make further progress without winetricks' vcrun201(7|9) requirement.

## Details
When tested Cxbx-Reloaded with Wine on Linux platform, it will crash every time on start of emulation due to `vcomp140.dll.C2VectParallel` is unimplemented. Wine 6.3 development build has contain a fix for this issue. It should resolve initial start of emulation process. However, it needs further testing to see any other crash may occur for any unimplemented functions. Or may not work as intended during emulation versus official Microsoft's redist binaries.

### System Configuration:
N/A

### Additional Information (if any):
references:
- [Wine 6.3 dev release news](https://www.winehq.org/announce/6.3)
- [Wine bug ticket for unimplemented C2VectParallel](https://bugs.winehq.org/show_bug.cgi?id=46238)

**Comment by lorn10** [comment]
Hi

This is my first noob contribution here. ;-)

Unfortunately, at least for current Wine development version 6.6 it seems that the mentioned newly implemented functions don't work. However, in my case Cxbx-Reloaded also doesn't work when vcrun2017 or vcrun2019 is installed. That may be most likely another bug of newer Wine 6.x releases...

The symptom is that when I double-click `cxbx.exe` the program doesn't start properly. It seems that it terminates right after it is started. When I try to run it from terminal (aka CLI console) then I will get a strange `c0000135` error which indicates that the path is wrong. Well, the path is 100% correct so I don't understand were is the problem.

```
test@test-iMac:~/Downloads$ WINEPREFIX=~/.wine wine /home/test/Downloads/Cxbx/2017/cxbx.exe
0034:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
0034:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
0034:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
0034:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
0034:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
0034:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
005c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
005c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
005c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
005c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
005c:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
005c:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
0064:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
0064:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
0064:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
0064:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
0064:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
0064:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
006c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
006c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
006c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
006c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
006c:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
006c:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
002c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
002c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
002c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
002c:fixme:font:get_name_record_codepage encoding 29 not handled, platform 1.
002c:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
002c:fixme:font:get_name_record_codepage encoding 20 not handled, platform 1.
wine: failed to open "/home/test/Downloads/Cxbx/2017/cxbx.exe": c0000135
```

**Comment by RadWolfie** [comment]
> Okay I am now on an absolute stock Wine 6.6 prefix and both Cxbx-Reloaded [f0e99ba](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/commit/f0e99ba436bb2d8b7dc6a4bd5575489e18d0d11d) builds are running. Absolutely nothing is installed with winetricks, also no d3dcompiler_47 dll file.

That's interesting, d3dcompiler_47 is needed during emulation with 3D environment. Perhaps somewhere in the process has implement for d3dcompiler support in Wine?

> However, the only strange thing is that Cxbx-Reloaded only starts from terminal. It doesn't starts when the `cxbx.exe` file is double-clicked. Will provide more information when I have looked somewhat deeper in that matter.

It may has to do with file allocation tie to certain programs in order to launch. If that's not the issue, then I wish you luck finding the cause of issue.

**Comment by lorn10** [comment]
Sorry, I couldn't test a game so far because I have at moment simply none available. ;-) My bro will bring me some of his Xbox classics sometimes soon (I hope) then I will test this out somewhat deeper. (The d3dcompiler_47.dll file my be needed when a game is loaded and some DX9 shader stuff is build.)

So for the moment, - Cxbx-Reloaded starts up and runs error free in a clean Wine 6.6 prefix. Will see how the games will load / work. And maybe I will also test this with Gallium Nine because most of my hardware is non-Vulkan based.

However, out of my perspective it is really great to see that Cxbx-Reloaded is starting absolutely fine under latest Wine. That cannot be taken for granted. For example, may other little "emulation testing project", PCSX-Redux (a total rewrite of the original PCSX emulator) is totally crashing in Wine because the "Windows Gaming Input" stuff is not fully available / implemented. ;-) It's amazing to see that CXBX-R still works despite that missing "Gaming Input support" in Wine.

![Screenshot_cxbxr_f0e99ba](https://user-images.githubusercontent.com/9931771/115683560-e0440080-a356-11eb-88a0-4212b99755b0.png)

**Comment by lorn10** [comment]
Here follows a new little update. At first, the `d3dcompiler_47.dll` file is definitively still needed. ;-) Despite that it unfortunately looks that CXBX-R is again broken on Wine 6.6. :-1: 

The newly implemented `C2VectParallel` function seems to be not the sole missing one, - when a game is started it simply doesn't load. The only thing which changes is that in the lower right corner of the emulator a picture of the game shows up, - in my case "Crash Tag Team Racing". This looks as follow:
![Screenshot_cxbxr_f0e99ba_inbuiltVS2019](https://user-images.githubusercontent.com/9931771/115993435-0a095b80-a5d3-11eb-9d97-dc04ecbcc31b.png)

However, when the latest MS Visual C++ 2019 libraries are installed with winetricks the emulator crashes right during a Xbe is loaded. As the error logs indicates, the problem has to do with the `ucrtbase` function. When this function is again set to _Wine builtin_, CXBX-R no longer crashes, it simply stands aging still at the above mentioned "game cover in the lower right corner" position.

[backtraceCXBXvs2019.txt](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/6371935/backtraceCXBXvs2019.txt)
[backtraceCXBXvs2019cli.txt](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/files/6371936/backtraceCXBXvs2019cli.txt)

**Comment by ergo720** [comment]
> However, when the latest MS Visual C++ 2019 libraries are installed with winetricks the emulator crashes right during a Xbe is loaded. 

I can confirm this behavior: I installed wine-staging 6.7 on Debian 10 (buster) running under VMware, and then installed vcrun2019 with winetricks, and as soon as you load an xbe in the gui, the program will crash.

@lorn10 However, these are really bugs in wine, and not of cxbxr, because we know it works just fine under Windows. Thus, you should report the bugs/logs to wine, not us, because, as far as we are concerned, we don't have anything to fix.

**Comment by lorn10** [comment]
Yes, also on Kubuntu 20.04 LTS and very latest Wine 6.7 devel CXBX-R is crashing when a Xbe is loaded. However, this is true only when the MS Visual C++ 2019 libraries are installed.

With the builtin libraries the situation look slightly better then on Wine 6.6 devel. When "start" is selected, the emulation process goes "one step further" beyond the "game picture in the lower right corner" stage. It still doesn't work, the emulation hangs shortly after at an empty (black) screen (within CXBX-R). But during this the emulation status sign (above, next to `LOG:`) turns green and before that I have a `FPS: 0.00 MS / F : 0.00` displayed. :-) Never get so far with Wine 6.6.

All this ends shortly thereafter and at the Linux CLI I get as last message: `wineserver crashed, please enable coredumps (ulimit -c unlimited) and restart.`

Whatever, it might get better in newer versions like Wine 6.8 or 7.0, we will see. But for now, the status in Wine seems again broken. This is most likely true for all 6.x releases in which the `C2VectParallel` fix is implemented.

**Comment by lorn10** [comment]
Okay folks, - it looks that CXBX-R (build cec1b0b) is **working** again! :+1:   The interesting thing is that it seems to work without any MS Visual C++ libraries installed. In fact, it works better when **no** original MS Visual C++ libraries are in use. 

I can confirm this for _Crash Tag Team Racing (EUR)_. CXBX-R is still crashing when the XBE is loaded and the MS Visual C++ libraries are used. When I set them back to _Builtin, Native_ then I am able to load the XBE and start the emulation. However, this is not true for _Baldur's Gate Dark Alliance (EUR)_, that XBE can be loaded regardless if the Visual C++ libraries are used or not, - weird. 

Well, there exist a little difference in the behavior between these two XBEs. While _Baldur's Gate Dark Alliance (EUR)_ is in perfect condition, no warning appears, there occurs one when _Crash Tag Team Racing (EUR)_ is loaded: `-XBE signature check failed!` 

It is really a mess that an original Xbox DVD cannot be loaded at a normal PC DVD drive. Whatever, @ergo720 It looks that a change in CXBX-R made it work again. So this wasn't a Wine issue after all... ;-)

Will look somewhat deeper into this matter when I find the time. And I will retest everything with a new, **clean** Wine 6.8 prefix. Here follows for the moment two screenshots:
![Screenshot_cxbxr_cec1b0b_Dark_Alliance](https://user-images.githubusercontent.com/9931771/118837347-006acf00-b8c5-11eb-9b66-a48ac2bf3aee.png)
![Screenshot_cxbxr_cec1b0b_Crash_Tag_Team_Racing](https://user-images.githubusercontent.com/9931771/118837375-0660b000-b8c5-11eb-9aa7-2830775760ea.png)

**Comment by lorn10** [comment]
Again a small addition. Also for the newly released **Wine 6.9**  development version it has to be said that the CXBX-R builds cec1b0b, 5d222c1 and df22c8b are **running perfectly**! :+1:  So this topic here can be marked as solved, or not @RadWolfie ?

My steps were identical as previously mentioned. In a new, clean Wine prefix I have just installed two components.

`winetricks -q d3dcompiler_47`

`winetricks -q galliumnine`

Well, as previously mentioned, Gallium Nine is not needed on modern Vulkan capable systems. But it makes a lot of sense on older, non-Vulkan compliant GPUs.

Final note regarding the binary building. Because newer Wine versions incorporate the `C2VectParallel` function, there are actually no longer two separate 2017 and 2019 versions needed. Or am I missing something, @LukeUsher ?

**Comment by PatrickvL** [comment]
Before we close this issue, the cxbx on wine installation description should be updated with above information

**Comment by RadWolfie** [comment]
> Again a small addition. Also for the newly released **Wine 6.9** development version it has to be said that the CXBX-R builds [cec1b0b](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/commit/cec1b0b8851756b3617a24bd51ff471466b12099), [5d222c1](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/commit/5d222c10a042ca1a342cd7e26af1c63624d65c97) and [df22c8b](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/commit/df22c8baedbafcbe3738b34f5845f416549aea6d) are **running perfectly**! 👍 So this topic here can be marked as solved, or not @RadWolfie ?

No, it cannot be mark as solved until Wine publish stable build containing the fix. It's been known over time their development builds break more things before we found out the issue is from Wine itself. It had been very tiresome from our side trying to figure out what was the cause.

> Before we close this issue, the cxbx on wine installation description should be updated with above information

Before we **even** close this issue. The fix must be upstream to stable build before we can update the readme information. 😉

**Comment by lorn10** [comment]
A new status update, - after even more intensive testing I have discovered an annoying little bug. Maybe that is related to a missing MS Visual C++ function / dll.

Regarding the above mentioned newer Wine builds **6.8** / **6.9** it is currently not possible to save in CXBX-R anything (from a running game) to the emulated internal "Xbox HDD". In the Linux terminal, I get the following message:
```
0160:fixme:file:NtCreateFile alloc_size not supported
0160:fixme:file:NtSetInformationFile Unsupported class (19)
0160:fixme:file:NtCreateFile alloc_size not supported
0160:fixme:file:NtSetInformationFile Unsupported class (19)
0160:fixme:file:NtCreateFile alloc_size not supported
0160:fixme:file:NtSetInformationFile Unsupported class (19)
0160:fixme:file:NtCreateFile alloc_size not supported
0160:fixme:file:NtSetInformationFile Unsupported class (19)
```
It looks that Wine simply doesn't implement the required `NtCreateFile alloc_size` and `NtSetInformationFile` functions. Can someone confirm that this works on the older stable Wine **5.0.3** release including the original MS Visual C++ Redistributable files?

Whatever, this seems now really to be a Wine limitation, that problem will be fixed when the Wine devs find the time to implement it.

**Comment by lorn10** [comment]
Here follows a short July Update regarding this topic.

The built-in Wine Visual C++ support is working stable and reliable also at the last Wine releases **6.11**, **6.12** and **6.13**.

This is also true for Wine 6.10 but this version had broken gamepad support. That happened because of a heavy refactoring and expansion of the Wine dinput codebase. Whatever, this was resolved again in Wine 6.11, the corresponding Wine dev Rémi Bernon was very cooperative and he fixed the underlying problem right after it was reported. :+1: 

So out of mine view the only remaining **BIG issue** is the `NtCreateFile alloc_size not supported` **Wine limitation**. This effectively prevents the saving of game data in most games while in others the saving works only partially (because an other function is used?).

Will open a bug report if the issue is also present in the upcoming Wine version. I know this is effectively a Wine problem but there may exist an easy workaround for that which could be implemented in CXBX-R. Maybe...

Alternatively a CXBX-R dev @RadWolfie can "ping" a Wine dev and place an informal feature request, "please add `NtCreateFile alloc_size` support". ;-)

**Comment by lorn10** [comment]
And now follows a short December update. Just for reference, I have opened at **Wine HQ** a bug report regrading the above mentioned `NtCreateFile alloc_size not supported` issue:

_CXBX-R Xbox emulator cannot save game data, error "NtCreateFile alloc_size not supported"_
_https://bugs.winehq.org/show_bug.cgi?id=52171_

Okay, this maybe somehow to early but I wish all folks here happy holidays and a good switch to the new year! Thanks to all for their awesome work in this matter. :+1:

**Comment by RadWolfie** [comment]
Hello guys, I recently notice Wine 7.0.0 is out and ran Jet Set Radio Future on vm. It appear to be able to run fine without VS2019 redist regardless different issue not relative to this ticket. Although, I would like other testers' feedback if we still need vcrun2019 or not. As we are getting closer to drop VS2017 support due to GitHub's CI system is about to drop windows-2016 containing VS2017.

We may will need to create a separate ticket for console / dump output to file still doesn't work at the moment.

**Comment by lorn10** [comment]
Yes, I can confirm that Wine's built-in VS2019 works for me really good in conjunction with CXBX-R.:+1: Especially when I compare this with the native MS VS2019 version, - the Wine implementation is now more stable. Besides that, it looks that later Wine versions have some strange flaws in conjunction with the original MS VS2019.

So I would clearly **not recommend** to use native MS VS2019 any longer with later Wine versions.:wink: 

But despite the fact that the built-in VS2019 runs fine, it must be noted that it is still not fully complete. For example, while it works with CXBX-R, it doesn't work with PCSX-Redux. For the later I had to copy the **[msvcp140_atomic_wait.dll](https://github.com/grumpycoders/pcsx-redux/issues/656#issuecomment-964170706)** file manually into the program folder to get it work. (After that, Wine's built-in VS2019 implementation runs also good with PCSX-Redux.)

As mentioned in my previous post the only effective stumbling block is out of mine view currently Wine bug 52171 which effectively prevents the saving of any game data in _Crash Tag Team Racing_ (and most likely also other games).

**Comment by lorn10** [comment]
Just for the record, - Wine developer @stefand provided a possible fix regarding the mentioned "game data saving problem" of CXBX-R in Wine which can be found [here](https://bugs.winehq.org/show_bug.cgi?id=52171). (Contrary to my assumption this seems to be after all CXBX-R and not Wine related.)

Whatever, it would be great if a CXBX-R dev like maybe @RadWolfie could look into that.:+1: Thanks!

**Comment by RadWolfie** [comment]
Wine had released stable 7.0 with a side-effect of game saves issue. Older Wine version no longer function with current build for some time without our awareness. Closing as resolved since we updated the readme in recent pull request #2334.

**Comment by lorn10** [comment]
A final feedback from my side. CXBX-R [CI-dc1f93b](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/releases/tag/CI-dc1f93b) works almost perfectly under **Wine 7.5** without any additional installation of native MS VS2019 or VS2022 files. :+1: 

So the only remaining issue is the saving problem under Wine of certain Xbox games mentioned in #2314.

**Comment by PeterD23** [comment]
@lorn10 
Just tried to run on the Steam Deck with Wine 7.0, and despite following the above steps I'm getting IDirect3D::CreateDeviceEx and unsure of how to proceed. It appears my wineprefix has all the correct dependencies, but it doesn't seem to recognise a valid driver being installed?

**Comment by stefand** [comment]
The steam deck might force dxvk's d3d9 implementation instead of wined3d on you. Try to switch between the two.

**Comment by LukeUsher** [comment]
It's important to note that Wine != Proton.

Steam's proton is *based on wine*, but it is a separate product with it's own versioning; Cxbx-Reloaded has only been tested on stock wine, it has not been tested on Valve's proton.

**Comment by PeterD23** [comment]
@stefand 
So I think DXVK doesn't work with CXBX-R. I wiped my wineprefix clean and tried doing just installing d3dcompiler_47 and galliumnine and it appears to work perfectly now. Proton didn't work when I tried running it through Steam so I think that just confirms my theory.

**Comment by lorn10** [comment]
@PeterD23 
Yes, I use **[Gallium Nine](https://github.com/iXit/wine-nine-standalone)** not DXVK with stock **Wine** and not Proton. :wink: 

And of course, WineD3D also works, however some effects may look different.

@LukeUsher Maybe you should really add in the Prerequisites section an information about DXVK and Gallium Nine.
